app/form_objects/email_confirmation_form.rb

```diff
@@ -16,4 +16,12 @@ def validate_email?
   def validate_presence?
     FeatureService.enabled?(:email_confirmations_enabled)
   end
+
+  def generate_submission_references!
```

Comment:
```suggestion
	def initialize
	   generate_submission_references
	end
  
  private
  def generate_submission_references
```

I don't think we track the notify reference between page loads in this controller so it should be safe enough to generate a new reference every time.  I would make `generate_submission_references` a private method, move `initialize` method so its just after the validates section above in this file.

The benefit of doing this is just by calling `EmailConfirmationForm.new` should set the notify references and then the controller doesn't need to worry about calling it at a specific time like https://github.com/alphagov/forms-runner/pull/500/files#diff-05a8792312bd570f7c19639941639759b413f0c20ef61c4a10856e28dddc2822R8

Should be easy enough to test if you stub out SecureRandom and then call `EmailConfirmationForm.new` and expect the two notify references to be set to the values you used.

Reply:
What about when `submit_answers` happens? Will `EmailConfirmationForm.new(email_confirmation_form_params)` generate a new set of references? We need to use the same references as were generated for the `show` route.

Reply:
Good point, We could do something like 7e756e6c8bf300318aa3ea5ad75ff160b35c8191 or use a guard 1bd844a4ab0aa91733449e2cb29549bb82e2a2e6 so only generate those value if they are set when the object is initialised

Reply:
Ah, I forgot to mention, in the case where there is an error submitting answers, we then do want to generate a new set of references, even if we already have some. See  https://github.com/alphagov/forms-runner/blob/18c6550b170efca3283023893e907cbafbc09646/app/controllers/forms/check_your_answers_controller.rb#L29-L31

Reply:
TBH, I think the only reason this was originally added was because if there was validation error nothing would have generated the values again so the page would have rendered with empty fields not for any other particular reason.

I dont see there being any harm in removing this and playing back the same reference number if there is a validation error.

Reply:
> TBH, I think the only reason this was originally added was because if there was validation error nothing would have generated the values again so the page would have rendered with empty fields

That's not the behaviour I'm seeing... in the `main` branch the `email_confirmation_form` will get the value from the form data in the request, and if line 31 is commented out and the form is not valid the rendered form will include the same reference number.

@DavidBiddle it looks like you introduced this line in https://github.com/alphagov/forms-runner/commit/429d381c90b6d5e1cea2e28010c00cf80501079f, do you happen to remember the reason for it? I can't see anything mentioned in the commit message.

Reply:
> @DavidBiddle it looks like you introduced this line in https://github.com/alphagov/forms-runner/commit/429d381c90b6d5e1cea2e28010c00cf80501079f, do you happen to remember the reason for it?

I don't think it was a particularly considered choice on my part - I think I just wasn't 100% sure whether we needed a new UUID or could keep the old one, so opted to generate a fresh one to be on the safe side.

Reply:
for simplicity and consistency with how other form fields on the page work, we should keep the same generated uuid between page submits if there are validation errors.

Reply:
This has been changed in commit [98245df](https://github.com/alphagov/forms-runner/pull/500/commits/98245df3dfb4b9a9d60e1363e3e7de26d3a12301).